c++/src/kj/mutex.c++

```diff
@@ -688,6 +738,16 @@ void Mutex::lockWhen(Predicate& predicate, Maybe<Duration> timeout) {
     KJ_PTHREAD_CALL(pthread_mutex_unlock(&waiter.stupidMutex));
 
     lock(EXCLUSIVE);
+    currentlyLocked = true;
+
+    KJ_IF_MAYBE(exception, waiter.exception) {
```

Comment:
Should this go above the `lock()`, since the exception will cause the mutex to be unlocked anyway?

Reply:
It can't because the waiter is still in the linked list where it is protected by the lock. In fact it occurs to me that I need to be a little more careful that my calls to `removeWaiter()` happen strictly under lock -- added a commit to ensure that.

Reply:
Oof, subtle!